wal/src/message_queue_impl/namespace.rs

```diff
@@ -0,0 +1,714 @@
+        // If inner iter is `Some`, poll it to get required log entry.
+        if let Some(iter) = self.iter.as_mut() {
+            if self.is_terminated {
+                return Ok(None);
+            }
+
+            let start_sequence = self.start;
+            let end_sequence = self.end;
+            loop {
+                let poll_result = iter.next_log_entry().await.context(ReadWithCause {
+                    namespace: self.namespace.clone(),
+                    request: self.request.clone(),
+                    msg: "failed while polling log",
+                })?;
+
+                if let Some(log_entry) = poll_result {
+                    // While polling result is `Some`, if:
+                    //  + the table id equals to the one in `request`
+                    //  + the sequence number of log entry is among [start, end], return this log
+                    // entry. Otherwise, just continue return the `log_entry`.
+                    // Otherwise, just continue to poll.
+                    if log_entry.table_id == self.request.location.table_id
+                        && log_entry.sequence >= start_sequence
+                        && log_entry.sequence <= end_sequence
+                    {
+                        self.is_terminated = log_entry.sequence == self.end;
+                        self.previous_value = log_entry.payload.to_owned();
+
+                        return Ok(Some(LogEntry {
+                            table_id: log_entry.table_id,
+                            sequence: log_entry.sequence,
+                            payload: &self.previous_value,
+                        }));
+                    }
+                } else {
+                    // While polling result is `None`, it represents has finished to poll inner
+                    // iter.
+                    self.is_terminated = true;
+
+                    break;
+                }
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+/// Loop all [Region] in [Namespace] and clean deleted logs.
+async fn log_cleaner_routine<M: MessageQueue>(inner: Arc<NamespaceInner<M>>) {
+    debug!(
+        "Periodical log cleaning process start, namespace:{}",
+        inner.namespace,
+    );
+
+    if let Err(e) = inner.clean_logs().await {
+        error!(
+            "Failed to clean deleted logs, namespace:{}, err:{}",
+            inner.namespace, e,
+        );
+    }
+
+    debug!(
+        "Periodical log cleaning process end, namespace:{}",
+        inner.namespace,
+    );
+}
+
+fn start_log_cleaner<M: MessageQueue>(
+    runtime: &Runtime,
+    period: Duration,
+    inner: Arc<NamespaceInner<M>>,
+) -> TaskHandle {
+    let name = format!("Log-cleaner-wal-on-mq-{}", inner.namespace);
+    let builder = move || {
+        let inner = inner.clone();
+
+        log_cleaner_routine(inner)
+    };
+
+    TimedTask::start_timed_task(name, runtime, period, builder)
```

Comment:
I don't see when to stop such task.

Reply:
It will be stopped while `Namespace::close` is called. 
I guess we design the method `close_gracefully` in interface is for doing this, but unfortunately I found it is just dead code now. 
I will implement drop to calling make the routine stop temporarily, and see how to make use of `close_gracefully` later.